File: code/977.squares-of-a-sorted-array.java
/*
 * @lc app=leetcode id=977 lang=java
 *
 * [977] Squares of a Sorted Array
 */

// @lc code=start

// import java.util.PriorityQueue;

class Solution {
    public int[] sortedSquares(int[] nums) {
        int[] res = new int[nums.length];
        int i = nums.length - 1;
        int left = 0;
        int right = nums.length - 1;
        while (left <= right) {
            if (nums[left] * nums[left] < nums[right] * nums[right]) {
                res[i] = nums[right] * nums[right];
                right--;
                i--;
            } else {
                res[i] = nums[left] * nums[left];
                left++;
                i--;
            }
        }
        return res;
    }
}
// @lc code=end

File: code/977.squares-of-a-sorted-array.go
/*
 * @lc app=leetcode id=977 lang=golang
 *
 * [977] Squares of a Sorted Array
 */

// @lc code=start
func sortedSquares(nums []int) []int {
	left, right, i := 0, len(nums)-1, len(nums)-1
	res := make([]int, len(nums))

	for left <= right {
		squareL, squareR := nums[left]*nums[left], nums[right]*nums[right]
		if squareL < squareR {
			res[i] = squareR
			right--
		} else {
			res[i] = squareL
			left++
		}
		i--
	}
	return res
}

// @lc code=end

